core/src/main/java/org/apache/iceberg/io/FileIOCloser.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.io;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import java.io.Closeable;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+/**
+ * Keeps track of the {@link FileIO} instance of the given {@link TableOperations} instance and
+ * closes the {@link FileIO} when {@link FileIOCloser#close()} gets called
+ */
+public class FileIOCloser implements Closeable {
```

Comment:
I feel like a better name is `FileIOTracker`? Yes it does achieve the goal of closing it when this is closed but really the direct interface most people would use would be the `trackFileIO` API? The name keeps it a bit more open in case there's something else folks want to extend and do something else to these FileIO's beyond closing

Reply:
I don't have strong opinions on the naming, so calling it `FileIOTracker` is fine IMO